MobiusCore/Source/MobiusController.swift

```diff
@@ -199,13 +168,152 @@ public final class MobiusController<Model, Event, Effect> {
+    // large models.
+    //
+    // `running` can safely be read from any thread. Changing `running` and reading or writing state is restricted
+    // to the loop queue.
+    private final class State {
+        // Intermediate states are required to handle the fun, fun case where an event source dispatches an event
+        // immediately, which happens before the loop variable is assigned in start(). In this case, we enter
+        // flipEventsToLoopQueue() in the .transitioningToRunning state, but the async block cannot proceed until we
+        // reach the .running state (because start() is already running on the loop queue).
+        private enum RawState {
+            case stopped
+            case transitioningToRunning
+            case running
+            case transitioningToStopped
+        }
+
+        private let rawState = Synchronized(value: RawState.stopped)
+        private let loopQueue: DispatchQueue
+        private var stoppedState: StoppedState?
+        private var runningState: RunningState?
+
+        enum Snapshot {
+            case stopped(StoppedState)
+            case running(RunningState)
+        }
+
+        init(model: Model, queue: DispatchQueue) {
+            stoppedState = StoppedState(modelToStartForm: model, viewConnectable: nil)
+            loopQueue = queue
+        }
+
+        var running: Bool {
+            switch rawState.value {
+            case .stopped:
+                return false
+            case .transitioningToRunning, .running, .transitioningToStopped:
+                return true
+            }
+        }
+
+        func syncRead<T>(_ closure: (Snapshot) throws -> T) rethrows -> T {
+            return try loopQueue.sync {
+                try closure(snapshot())
+            }
+        }
+
+        func syncMutateStopped(error: String, _ closure: (inout StoppedState) -> Void) {
+            dispatchPrecondition(condition: .notOnQueue(loopQueue))
+            loopQueue.sync {
+                switch snapshot() {
+                case .running:
+                    MobiusHooks.onError(error)
+                case .stopped(var state):
+                    closure(&state)
+                    stoppedState = state
+                }
+            }
+        }
+
+        func syncTransitionToRunning(error: String, _ transition: (StoppedState) -> RunningState?) {
+            dispatchPrecondition(condition: .notOnQueue(loopQueue))
+            loopQueue.sync {
+                switch snapshot() {
+                case .running:
+                    MobiusHooks.onError(error)
+                case .stopped(let stoppedState):
+                    rawState.value = .transitioningToRunning
+                    if let runningState = transition(stoppedState) {
+                        become(running: runningState)
+                    } else {
+                        rawState.value = .stopped
```

Comment:
I’ll add a doc comment. This happens if the transition function returns nil, which happens after an `onError` (since the error hook is non-terminating).